### PR TITLE
TEST: PL_open_query with various options plus some foreign.doc examples

### PR DESCRIPTION
**Pull request by kamahen**
These tests are to help verify my understanding of foreign.doc (including some changes I made in https://github.com/SWI-Prolog/swipl-devel/pull/1119

### CODE REVIEW

File: test_ffi.c [hunk 1]
```diff
@@ -187,83 +181,310 @@ ffi_options_(term_t a1, term_t options)
...
+  return PL_unify(l, env);
+}
 
-  return TRUE;
+foreign_t
+ffi_write_atoms_(term_t stream, term_t l)
+{ term_t head = PL_new_term_ref();   /* the elements */
+  term_t list = PL_copy_term_ref(l); /* copy (we modify list) */
+  IOSTREAM* s;
+  if ( !PL_get_stream(stream, &s, SIO_OUTPUT) )
```
Comment (JanWielemaker) [inline test_ffi.c:447]: This needs a matching PL_release_stream().   Otherwise you may deadlock

**Comment by kamahen** [comment]
The latest commit matches the documentation changes in https://github.com/SWI-Prolog/swipl-devel/pull/1122

File: test_ffi.pl [hunk 2]
```diff
@@ -319,3 +320,76 @@
     ffi_findall_bug_(ffi_p(_), L).
 
 :- end_tests(call).
+
```
Comment (JanWielemaker) [inline test_ffi.pl]: This should go into src/Tests/core/test_text.pl

**Review by JanWielemaker — CHANGES_REQUESTED** [review]
Oops, see I didn't finish my review yesterday.    Please find comment on stream handles and sub_atom/string tests.   Tests are great.  Better put them at the right place immediately though :smile:

File: test_ffi.c (hunk 1 above)
Comment (kamahen) [reply test_ffi.c:447]: Also changed sample code in foreign.doc

File: test_ffi.pl (hunk 2 above)
Comment (kamahen) [reply test_ffi.pl]: Done

**Comment by JanWielemaker** [comment]
Merged after some edit and squash.    Fixed one of your Unicode TODO.   In general I consider the atom/string <-> char*/wchar_t* a bit old school.   The PL_{get_unify_put}_[nw]chars() family is more friendly when it comes to different input, output, encoding and exception handling.

**Pull request closed without merging**